dask/highlevelgraph.py

```diff
@@ -24,6 +27,122 @@ def _find_layer_containing_key(key):
     return ret
 
 
+class Layer(Mapping):
+    """High level graph layer
+
+    This abstract class establish a protocol for high level graph layers.
+    """
+
+    def cull(self, keys: Set) -> "Layer":
+        """Return a new Layer with only the tasks required to calculate `keys`.
+
+        In other words, remove unnecessary tasks from the layer.
+
+        Examples
+        --------
+        >>> d = Layer({'x': 1, 'y': (inc, 'x'), 'out': (add, 'x', 10)})  # doctest: +SKIP
+        >>> d.cull({'out'})  # doctest: +SKIP
+        {'x': 1, 'out': (add, 'x', 10)}
+
+        Returns
+        -------
+        layer: Layer
+            Culled layer
+        """
+
+        seen = set()
+        out = {}
+        work = set(flatten(keys))
+        while work:
+            new_work = set()
+            for k in work:
+                out[k] = self[k]
+                for d in get_dependencies(self, k):
+                    if d not in seen:
+                        seen.add(d)
+                        new_work.add(d)
+            work = new_work
+
+        return BasicLayer(out)
+
+    def get_external_dependencies(self, all_hlg_keys: Set) -> Set:
+        """Get external dependencies
+
+        Parameters
+        ----------
+        all_hlg_keys : set
+            Set of all keys in the high level graph.
+
+        Returns
+        -------
+        deps: set
+            Set of dependencies
+        """
+
+        all_deps = keys_in_tasks(all_hlg_keys, self.values())
+        return all_deps.difference(self.keys())
+
+
+class BasicLayer(Layer):
+    """Basic implementation of `Layer`
+
+    Parameters
+    ----------
+    mapping : Mapping
+        The mapping between keys and tasks, typically a dask graph.
+    key_dependencies : Mapping[Hashable, Set], optional
+        Mapping between keys and their dependencies
```

Comment:
I have a slight preference to call this just `dependencies`.  I understand that that term is over-used, but I don't think that prefixing with `key_` significantly improves things.  Accessing `.dependencies` is a pretty common thing to do in a variety of Dask objects.

Reply:
Fixed